boot/seal.go

```diff
@@ -517,57 +526,86 @@ func resealFallbackObjectKeys(pbc predictableBootChains, authKeyFile string, rol
 
 // TODO:UC20: this needs to take more than one model to accommodate the remodel
 // scenario
-func recoveryBootChainsForSystems(systems []string, trbl bootloader.TrustedAssetsBootloader, modeenv *Modeenv) (chains []bootChain, err error) {
-	model := modeenv.ModelForSealing()
-	for _, system := range systems {
-		// get kernel and gadget information from seed
-		perf := timings.New(nil)
-		_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, system, []snap.Type{snap.TypeKernel, snap.TypeGadget}, perf)
-		if err != nil {
-			return nil, fmt.Errorf("cannot read system %q seed: %v", system, err)
-		}
-		if len(snaps) != 2 {
-			return nil, fmt.Errorf("cannot obtain recovery system snaps")
-		}
-		seedKernel, seedGadget := snaps[0], snaps[1]
-		if snaps[0].EssentialType == snap.TypeGadget {
-			seedKernel, seedGadget = seedGadget, seedKernel
-		}
+func recoveryBootChainsForSystems(systems []string, trbl bootloader.TrustedAssetsBootloader, modeenv *Modeenv, includeTryModel bool) (chains []bootChain, err error) {
+
+	chainsForModel := func(model secboot.ModelForSealing, allowedModels []string) error {
+		for _, system := range systems {
+			// get kernel and gadget information from seed
+			perf := timings.New(nil)
+			seedSystemModel, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, system, []snap.Type{snap.TypeKernel, snap.TypeGadget}, perf)
+			if err != nil {
+				return fmt.Errorf("cannot read system %q seed: %v", system, err)
+			}
+			if len(snaps) != 2 {
+				return fmt.Errorf("cannot obtain recovery system snaps")
+			}
+			seedModelID := modelUniqueID(seedSystemModel)
+			if !strutil.ListContains(allowedModels, seedModelID) {
```

Comment:
I don't understand why we need allowedModels,  I think here we just need to do check:

`modelUniqueID(model) != seedModelID`

instead, no?

each of the recovery systems has exactly one model

Reply:
Yes, I think it should be ok. Trying to follow this path again, the fallback key is sealed to specific recovery systems, and is tried in recover mode, when the run key did not work. In this scenario snap-bootstrap measures the model of a particular recovery system, which the fallback key should have been resealed for right here. The recovery system always has just one models, so checking against a list of models is indeed unnecessary.

Reply:
Updated, I've also added more comments to the unit test explaining what is happening there.